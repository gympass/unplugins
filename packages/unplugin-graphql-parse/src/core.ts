import { parse } from 'graphql';

import type { Options } from './options';

export function transform(code: string) {
  const schema = JSON.stringify(parse(code));

  const contents = `const schema = ${schema};
export default schema;
`;

  return contents;
}

export function transformInclude(id: string, options: Options) {
  return id.endsWith(options.ext);
}
