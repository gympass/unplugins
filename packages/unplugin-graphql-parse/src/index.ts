import gql from 'graphql';
import { createUnplugin } from 'unplugin';

export type Options = {
  ext: string;
};

const defaults: Options = { ext: '.graphql' };

const unpluginGraphQLParser = createUnplugin<Options | undefined>(
  (options = defaults) => ({
    name: 'unplugin-graphql-parser',
    transformInclude(id) {
      return id.endsWith(options.ext);
    },
    transform(code) {
      const schema = JSON.stringify(gql.parse(code));

      const contents = `const schema = ${schema};
export default schema;
`;

      return contents;
    },
  }),
);

export default unpluginGraphQLParser;
