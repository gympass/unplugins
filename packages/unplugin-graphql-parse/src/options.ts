export type Options = {
  ext: string;
};
