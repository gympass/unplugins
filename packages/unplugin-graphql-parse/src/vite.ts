import unplugin from './index';
import type { Options } from './index';

export default unplugin.vite as (options?: Options) => any;
