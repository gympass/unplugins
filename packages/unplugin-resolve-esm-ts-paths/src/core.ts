import { parse } from 'es-module-lexer';

import { dirname } from 'node:path';

import type { Options } from './options.js';
import { resolveGlob } from './resolvers/glob.js';
import { resolveRelative } from './resolvers/relative.js';
import { resolvePaths } from './resolvers/paths.js';
import {
  BROKEN_IMPORT,
  MODULE_IMPORT,
  escapePathForRegExp,
} from './helpers/index.js';

export async function transform(code: string, id: string, options: Options) {
  const dir = dirname(id);

  const endExt = options.resolvedExt;

  try {
    const [imports] = await parse(code);

    if (imports.length === 0) return code;

    const replaceMapPromises = imports.map<Promise<[string, string]>>(
      fileImport => {
        const file = fileImport.n;
        if (!file) return Promise.resolve([BROKEN_IMPORT, BROKEN_IMPORT]);

        const importString = code.substring(fileImport.ss, fileImport.se);

        const correctQuote = (
          /'|"/.exec(importString) as RegExpExecArray
        ).pop() as string;

        // Resolve Glob imports.
        if (options.resolveGlob && file.includes('glob:')) {
          return resolveGlob(id, fileImport, dir, endExt, importString);
        }

        // Add index file import for relative directory imports.
        if (file.includes('./')) {
          return resolveRelative(file, dir, endExt, correctQuote);
        }

        return resolvePaths(file, dir, endExt, correctQuote);
      },
    );

    // Clear the map keeping unique keys and remove module imports.
    const replaceMap = new Map<string, string>(
      await Promise.all(replaceMapPromises),
    );
    replaceMap.delete(MODULE_IMPORT);
    replaceMap.delete(BROKEN_IMPORT);

    // If size is zero, there are only modules imports.
    if (replaceMap.size === 0) return code;

    const allPathsRegExp = new RegExp(
      [...replaceMap.keys()].map(escapePathForRegExp).join('|'),
      'g',
    );

    return code.replace(allPathsRegExp, match => {
      return replaceMap.get(match) as string;
    });
  } catch (error) {
    /**
     * If we get a problem during parsing, we return the original
     * code so the bundler can handle it.
     */
    return code;
  }
}

export function transformInclude(id: string, options: Options) {
  if (options.ext instanceof RegExp) {
    return Boolean(id.match(options.ext));
  }
  return id.endsWith(options.ext);
}
