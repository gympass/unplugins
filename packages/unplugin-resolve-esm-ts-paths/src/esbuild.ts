import unplugin from './index.js';

export default unplugin.esbuild;
