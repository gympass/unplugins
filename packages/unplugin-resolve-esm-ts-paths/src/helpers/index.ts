import { stat } from 'node:fs/promises';

export const BROKEN_IMPORT = 'BROKEN_IMPORT';
export const MODULE_IMPORT = 'MODULE_IMPORT';

export async function isDirectory(path: string) {
  try {
    return (await stat(path)).isDirectory();
  } catch (e) {
    return false;
  }
}

export function escapePathForRegExp(path: string) {
  return path.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
}

export function removeExt(path: string) {
  let pathWithoutDots = path;
  if (path.includes('..')) {
    const parts = path.split('..');
    pathWithoutDots = parts[parts.length - 1];
  }

  // https://stackoverflow.com/a/12900504/4223135
  const ext = pathWithoutDots.slice(
    // eslint-disable-next-line no-bitwise
    ((pathWithoutDots.lastIndexOf('.') - 1) >>> 0) + 2,
  );

  return path.replace(new RegExp(`\\.${ext}$`), '');
}
