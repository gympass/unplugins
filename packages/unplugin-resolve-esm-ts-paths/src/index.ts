import { createUnplugin } from 'unplugin';

import type { Options } from './options.js';
import { transform, transformInclude } from './core.js';

const defaults: Options = {
  ext: '.ts',
  resolveGlob: false,
  resolvedExt: '.js',
};

const unpluginResolveEsmTSPaths = createUnplugin<Options | undefined>(
  options => {
    const mergedDefaults = { ...defaults, ...options };

    return {
      name: 'unplugin-resolve-esm-ts-paths',
      transformInclude: id => transformInclude(id, mergedDefaults),
      transform: (code, id) => transform(code, id, mergedDefaults),
    };
  },
);

export default unpluginResolveEsmTSPaths;
