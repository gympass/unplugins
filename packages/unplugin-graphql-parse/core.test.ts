import { transform, transformInclude } from './src/core';

it('should accept the incomming ext', () => {
  expect(transformInclude('foo.graphql', { ext: '.graphql' })).toBe(true);
});

it('should parse graphql sdls', () => {
  const sdl = /* GraphQL */ `
    type MyType {
      id: ID
    }
  `;

  expect(transform(sdl)).toMatchSnapshot();
});
