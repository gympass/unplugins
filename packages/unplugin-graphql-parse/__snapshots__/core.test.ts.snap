// Vitest Snapshot v1

exports[`should parse graphql sdls 1`] = `
"const schema = {\\"kind\\":\\"Document\\",\\"definitions\\":[{\\"kind\\":\\"ObjectTypeDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"MyType\\",\\"loc\\":{\\"start\\":10,\\"end\\":16}},\\"interfaces\\":[],\\"directives\\":[],\\"fields\\":[{\\"kind\\":\\"FieldDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"id\\",\\"loc\\":{\\"start\\":25,\\"end\\":27}},\\"arguments\\":[],\\"type\\":{\\"kind\\":\\"NamedType\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"ID\\",\\"loc\\":{\\"start\\":29,\\"end\\":31}},\\"loc\\":{\\"start\\":29,\\"end\\":31}},\\"directives\\":[],\\"loc\\":{\\"start\\":25,\\"end\\":31}}],\\"loc\\":{\\"start\\":5,\\"end\\":37}}],\\"loc\\":{\\"start\\":0,\\"end\\":40}};
export default schema;
"
`;
